index.bs

```diff
@@ -3530,6 +3530,26 @@ operation. Likewise, clients can choose to produce a [=client extension output=]
 understand by encoding the [=authenticator extension output=] value into JSON, provided that the CBOR output uses only types
 present in JSON.
 
+When clients choose to pass through extensions they do not recognize,
+the JavaScript values in the [=client extension input=] values
+are converted to [=CBOR=] values in the [=authenticator extension input=] values.
+When the JavaScript input value is an [=%ArrayBuffer%=], it is converted to a [=CBOR=] byte array.
+When the JavaScript input value is a non-integer number, it is converted to a 64-bit CBOR floating point number.
```

Comment:
I thought that floats were banned in CTAP2 CBOR? Even if not, I would tend to ban them here:

Consider something like the geolocation extension. If it wished to send floats then it would need to ensure that the values in Javascript were never exact integers. Otherwise, as one moves across a line of latitude / longitude, the type signature of the CBOR would change. That seems fragile.

Rather, I would suggest that values are always rounded and that extensions use a fixed-point representation for fractional values.

Reply:
Issue https://github.com/fido-alliance/fido-2-specs/issues/419 is exactly about making sure that floating point values can be present in CTAP2 canonical CBOR - motivated in part by the geolocation extension.

The geolocation is a test case that we want to work.  Rather than banning floats, we should define that they are doubles unless integers - as the text already does.  It will be infinitesimally rare that longitude and latitude values are integers, so it will nearly always work.  Rounding longitude/latitude values to integers introduces errors of more than 60 nautical miles - a non-starter!

Reply:
https://github.com/fido-alliance/fido-2-specs/pull/456 defines that floats are allowed by CTAP2 canonicalization (and are not changed by it).

Reply:
If we go this route, the type-signature of extensions will fluctuate depending on the values involved, i.e. from float to int when crossing a line of latitude / longitude. This seems like a pool of bugs waiting to happen and it would be more stable to translate Javascript numbers as CBOR floats in all cases if we wish to have a generic transformation defined.

Also, the geolocation extension doesn't need floats. As noted above, they could just use fixed-point integers. (I.e. express a latitude as the integer round(x * 1000000), for whatever scale-factor works in their context.)

Reply:
The geolocation extension uses the Coordinates type defined at https://www.w3.org/TR/geolocation-API/#coordinates_interface , which does use floats (doubles).  It would be silly/unhelpful to define a different equivalent type in a W3C spec.

The generic conversion will always involve some tradeoffs.  Given that most numbers used in these interfaces are integers, representing them as such is certainly the right thing to do.  Prohibiting floats would be wrong/unhelpful too.

So yes, the generic conversion has tradeoffs, but I think we've written down the right ones for WebAuthn and CTAP2.

Reply:
It would not be silly to define the geolocation extension in terms of a scaled version of that structure in order to eliminate all the complexity of floats throughout this spec, I feel.

In fact, I think this has convinced me more strongly that the generic transformation is a mistake in general. I'll write a more general note about that below.